File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int N = Integer.parseInt(br.readLine());

        while (N-- > 0) {
            String[] str = br.readLine().split("");
            Stack<String> stack = new Stack<>();
            for (int i = 0; i < str.length; i++) {
                if (str[i].equals("(")) {
                    stack.push(str[i]);
                } else {
                    if (str[i].equals(")") && stack.contains("(")) {
//                        stack.remove("(");
                        stack.pop();
                    } else {
                        stack.push(str[i]);
                    }
                }
            }
            if (stack.isEmpty()) {
                sb.append("YES").append("\n");
            } else {
                sb.append("NO").append("\n");
            }

        }


        System.out.println(sb);

    }

}

File: 백준/Silver/9012. 괄호/괄호.py
from sys import stdin


N = int(stdin.readline().strip())

for _ in range(N):
    stack = []
    vps = stdin.readline().strip()
    for i in range(len(vps)):
        char = vps[i]
        if char == "(":
            stack.append(char)

        else:
            if len(stack) != 0 and stack[-1] == "(":
                stack.pop()
            else:
                stack.append(char)

    print("YES") if len(stack) == 0 else print("NO")

